ompi/mca/coll/base/coll_base_util.c

```diff
@@ -255,7 +255,7 @@ static int free_vecs_callback(struct ompi_request_t **rptr) {
 }
 
 int ompi_coll_base_retain_datatypes_w( ompi_request_t *req,
-                                       ompi_datatype_t *stypes[], ompi_datatype_t *rtypes[]) {
+                                       ompi_datatype_t * const stypes[], ompi_datatype_t * const rtypes[]) {
```

Comment:
Are you sure that this is correct -- add `const` to the param type in the declaration, only to cast it away in its usage in the definition?

Is there a reason to do that?  It seems like a no-op...?

Reply:
It is a no-op in this case. It's essentially a hack to make gcc happy. At the highest level API we're passing in a `const MPI_Datatype[]`, and somewhere down the stack (I suppose here), that const needs to be lost somewhere. I didn't want to touch the higher level api, so I figured here is the best place to fix it.

The gcc warning itself was converting a const pointer to a non-const pointer when using this function.

Reply:
@bosilca Can you chime in on this?

Reply:
Yeah, this warning was there for quite some time, but I was not sure if it is worth fixing it. We are altering the datatype inside the ```ompi_coll_base_retain_datatypes_w``` function (by increasing the refcount), so the datatype is not const, but the pointer to it should be. Thus, I think the proposed fix is correct.

However, we do need to cast it away below, as we cant change the type of ```vecs.stypes```and ```vecs.rtypes``` because these will be used later by the ```release_datatype``` function, where nothing shall be constant (the types can disappear if the user called ```MPI_Type_free``` on them).

Reply:
Yeah, that is correct. I tried changing the types of vect.types to also be constant and not have the cast. That not working, I settled for the current "fix".